driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -153,119 +155,165 @@ private static String extractMessage(Throwable t) {
+        return new AsyncFunction<Message.Response, Void>() {
+            @Override
+            public ListenableFuture<Void> apply(Message.Response response) throws Exception {
+                switch (response.type) {
+                    case READY:
+                        return checkClusterName(executor);
+                    case ERROR:
+                        Responses.Error error = (Responses.Error)response;
+                        // Testing for a specific string is a tad fragile but well, we don't have much choice
+                        if (error.code == ExceptionCode.PROTOCOL_ERROR && error.message.contains("Invalid or unsupported protocol version"))
+                            throw unsupportedProtocolVersionException(protocolVersion);
+                        throw defunct(new TransportException(address, String.format("Error initializing connection: %s", error.message)));
+                    case AUTHENTICATE:
+                        Authenticator authenticator = factory.authProvider.newAuthenticator(address);
+                        if (protocolVersion == 1)
+                        {
+                            if (authenticator instanceof ProtocolV1Authenticator)
+                                return authenticateV1(authenticator, executor);
+                            else
+                                // DSE 3.x always uses SASL authentication backported from protocol v2
+                                return authenticateV2(authenticator, executor);
+                        }
+                        else
+                            return authenticateV2(authenticator, executor);
+                    default:
+                        throw defunct(new TransportException(address, String.format("Unexpected %s response message from server to a STARTUP message", response.type)));
+                }
+            }
+        };
     }
 
-    private UnsupportedProtocolVersionException unsupportedProtocolVersionException(int triedVersion) {
-        logger.debug("Got unsupported protocol version error from {} for version {}", address, triedVersion);
-        return new UnsupportedProtocolVersionException(address, triedVersion);
+    // Due to C* gossip bugs, system.peers may report nodes that are gone from the cluster.
+    // If these nodes have been recommissionned to another cluster and are up, nothing prevents the driver from connecting
+    // to them. So we check that the cluster the node thinks it belongs to is our cluster (JAVA-397).
+    private ListenableFuture<Void> checkClusterName(final ListeningExecutorService executor) {
+        final String expected = factory.manager.metadata.clusterName;
+
+        // At initialization, the cluster is not known yet
+        if (expected == null)
+            return MoreFutures.VOID_SUCCESS;
+
+        DefaultResultSetFuture clusterNameFuture = new DefaultResultSetFuture(null,new Requests.Query("select cluster_name from system.local"));
```

Comment:
I agree with your analysis and fix. The potential NPE was already there before, but creating and closing the connections in parallel makes it much more likely to happen.